### PR TITLE
Add canonicalization for PyIR.PyFuncOp

### PR DESCRIPTION
**Pull request by AlexanderKalistratov**
Add canonicalization for PyIR.PyFuncOp which eliminates unused captures.

### CODE REVIEW

File: newlang/hckernel/mlir/lib/Dialect/PyIR/IR/PyIROps.cpp [hunk 1]
```diff
@@ -28,6 +28,61 @@ void hc::py_ir::PyIRDialect::initialize() {
...
+    if (!hasUnused)
+      return mlir::failure();
+
+    mlir::SmallVector<unsigned> indexes;
+    for (int i = captureNames.size() - 1; i >= 0; --i) {
+      auto capt = capturesBlockArgs[i];
+      if (capt.use_empty()) {
+        indexes.push_back(i);
+      }
+    }
```
Comment (Hardcode84) [inline newlang/hckernel/mlir/lib/Dialect/PyIR/IR/PyIROps.cpp]: You can unify these 2 loops and just check `indexes.empty()`

File: newlang/hckernel/mlir/lib/Dialect/PyIR/IR/PyIROps.cpp [hunk 2]
```diff
@@ -28,6 +28,61 @@ void hc::py_ir::PyIRDialect::initialize() {
...
+
+    if (!hasUnused)
+      return mlir::failure();
+
+    mlir::SmallVector<unsigned> indexes;
+    for (int i = captureNames.size() - 1; i >= 0; --i) {
+      auto capt = capturesBlockArgs[i];
+      if (capt.use_empty()) {
```
Comment (Hardcode84) [inline newlang/hckernel/mlir/lib/Dialect/PyIR/IR/PyIROps.cpp]: remove braces for single-line ifs

File: newlang/hckernel/mlir/lib/Dialect/PyIR/IR/PyIROps.cpp [hunk 3]
```diff
@@ -28,6 +28,61 @@ void hc::py_ir::PyIRDialect::initialize() {
...
+          op.getCaptureNames().getAsRange<mlir::Attribute>());
+      captNames.erase(captNames.begin() + idx);
+      auto attr =
+          rewriter.getArrayAttr(mlir::ArrayRef<mlir::Attribute>(captNames));
+      op.setCaptureNamesAttr(attr);
+      entryBlock->eraseArgument(blockArgIdx);
+    };
+
+    for (auto &&idx : indexes) {
+      eraseCaptureByIndex(idx);
```
Comment (Hardcode84) [inline newlang/hckernel/mlir/lib/Dialect/PyIR/IR/PyIROps.cpp]: You need to wrap inplace modifications in `modifyOpInPlace`

File: newlang/hckernel/tests/Dialect/PyIR/canonicalization.mlir [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+  %3 = typing.make_ident "i16" []
+  %4 = py_ir.func "func" (a:%0, b:%0) : !py_ir.undefined, !py_ir.undefined capture (c:%1, d:%2, e:%3) : !typing.value, !typing.value, !typing.value -> !py_ir.undefined {
+  ^bb0(%arg0: !py_ir.undefined, %arg1: !py_ir.undefined, %arg2: !py_ir.undefined, %arg3: !py_ir.undefined, %arg4: !py_ir.undefined):
+    %5 = py_ir.tuple_pack %arg0, %arg1 : !py_ir.undefined, !py_ir.undefined -> !py_ir.undefined
+    py_ir.return %5 : !py_ir.undefined
+  }
+  py_ir.module_end %4 : !py_ir.undefined
+}
+
+// -----
```
Comment (Hardcode84) [inline newlang/hckernel/tests/Dialect/PyIR/canonicalization.mlir]: separator is not needed at the end

File: newlang/hckernel/mlir/lib/Dialect/PyIR/IR/PyIROps.cpp [hunk 5]
```diff
@@ -143,6 +198,11 @@ static void printArgList(mlir::OpAsmPrinter &printer, Op /*op*/,
   printer << ')';
 }
 
+void hc::py_ir::PyFuncOp::getCanonicalizationPatterns(
```
Comment (Hardcode84) [inline newlang/hckernel/mlir/lib/Dialect/PyIR/IR/PyIROps.cpp]: move near to the rest of `PyFuncOp` funcs

File: newlang/hckernel/mlir/lib/Dialect/PyIR/IR/PyIROps.cpp [hunk 6]
```diff
@@ -28,6 +28,61 @@ void hc::py_ir::PyIRDialect::initialize() {
...
+      int blockArgIdx = idx + op.getBlockArgs().size();
+      auto *entryBlock = op.getEntryBlock();
+      op.getCaptureArgsMutable().erase(idx);
+      auto captNames = mlir::SmallVector<mlir::Attribute>(
+          op.getCaptureNames().getAsRange<mlir::Attribute>());
+      captNames.erase(captNames.begin() + idx);
+      auto attr =
+          rewriter.getArrayAttr(mlir::ArrayRef<mlir::Attribute>(captNames));
+      op.setCaptureNamesAttr(attr);
+      entryBlock->eraseArgument(blockArgIdx);
```
Comment (Hardcode84) [inline newlang/hckernel/mlir/lib/Dialect/PyIR/IR/PyIROps.cpp]: block has `eraseArguments` which takes a bitvector and allows a bulk erase.

File: newlang/hckernel/mlir/lib/Dialect/PyIR/IR/PyIROps.cpp (hunk 1 above)
Comment (AlexanderKalistratov) [reply newlang/hckernel/mlir/lib/Dialect/PyIR/IR/PyIROps.cpp]: I can, but I don't want

File: newlang/hckernel/mlir/lib/Dialect/PyIR/IR/PyIROps.cpp (hunk 2 above)
Comment (AlexanderKalistratov) [reply newlang/hckernel/mlir/lib/Dialect/PyIR/IR/PyIROps.cpp]: done

File: newlang/hckernel/mlir/lib/Dialect/PyIR/IR/PyIROps.cpp (hunk 3 above)
Comment (AlexanderKalistratov) [reply newlang/hckernel/mlir/lib/Dialect/PyIR/IR/PyIROps.cpp]: done

File: newlang/hckernel/tests/Dialect/PyIR/canonicalization.mlir (hunk 4 above)
Comment (AlexanderKalistratov) [reply newlang/hckernel/tests/Dialect/PyIR/canonicalization.mlir]: done

File: newlang/hckernel/mlir/lib/Dialect/PyIR/IR/PyIROps.cpp (hunk 5 above)
Comment (AlexanderKalistratov) [reply newlang/hckernel/mlir/lib/Dialect/PyIR/IR/PyIROps.cpp]: done

File: newlang/hckernel/mlir/lib/Dialect/PyIR/IR/PyIROps.cpp (hunk 6 above)
Comment (AlexanderKalistratov) [reply newlang/hckernel/mlir/lib/Dialect/PyIR/IR/PyIROps.cpp]: done

File: newlang/hckernel/mlir/lib/Dialect/PyIR/IR/PyIROps.cpp [hunk 7]
```diff
@@ -28,6 +28,62 @@ void hc::py_ir::PyIRDialect::initialize() {
...
+    for (int i = capturesBlockArgs.size() - 1; i >= 0; --i) {
+      auto capt = capturesBlockArgs[i];
+      if (capt.use_empty())
+        indexes.push_back(i);
+    }
+
+    rewriter.modifyOpInPlace(op, [&] {
+      auto *entryBlock = op.getEntryBlock();
+      auto allArgs = entryBlock->getArguments().size();
+      auto captNames = mlir::SmallVector<mlir::Attribute>(
```
Comment (Hardcode84) [inline newlang/hckernel/mlir/lib/Dialect/PyIR/IR/PyIROps.cpp]: `llvm::to_vector`

Comment (AlexanderKalistratov) [reply newlang/hckernel/mlir/lib/Dialect/PyIR/IR/PyIROps.cpp]: done

**Review by Hardcode84 — APPROVED** [review]

**Pull request merged**